src/Compilers/Core/Portable/Symbols/ISymbolExtensions_IsAccessibleWithin.cs

```diff
@@ -0,0 +1,353 @@
+                return true;
+            }
+
+            var originalThroughTypeOpt = throughTypeOpt == null ? null : throughTypeOpt.OriginalDefinition as ITypeSymbol;
+            for (INamedTypeSymbol current = withinType.OriginalDefinition; current != null; current = current.ContainingType)
+            {
+                if (current.InheritsFromIgnoringConstruction(originalContainingType))
+                {
+                    if (originalThroughTypeOpt == null ||
+                        originalThroughTypeOpt.InheritsFromIgnoringConstruction(current))
+                    {
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        private static bool IsPrivateSymbolAccessible(ISymbol within, INamedTypeSymbol containingType)
+        {
+            var withinType = within as INamedTypeSymbol;
+
+            // A private symbol is accessible if we're (optionally nested) inside the type that it
+            // was defined in.
+            return withinType?.IsNestedWithinOriginalContainingType(containingType.OriginalDefinition) == true;
+        }
+
+        private static bool IsNestedWithinOriginalContainingType(this INamedTypeSymbol type, INamedTypeSymbol possiblyContainingOriginalType)
+        {
+            Debug.Assert(possiblyContainingOriginalType.IsDefinition);
+            Debug.Assert(type != null);
+
+            for (var current = type.OriginalDefinition; current != null; current = current.ContainingType)
+            {
+                if (SameOriginalNamedType(current, possiblyContainingOriginalType))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private static bool InheritsFromIgnoringConstruction(this ITypeSymbol type, INamedTypeSymbol baseType)
+        {
+            Debug.Assert(type.IsDefinition);
+            Debug.Assert(baseType.IsDefinition);
+
+            for (var current = type; current != null; current = current.BaseType?.OriginalDefinition)
+            {
+                if (SameOriginalNamedType(baseType, current as INamedTypeSymbol))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private static bool SameOriginalNamedType(INamedTypeSymbol t1, INamedTypeSymbol t2)
+        {
+            Debug.Assert(t1 != null);
+            Debug.Assert(t1.IsDefinition);
+            Debug.Assert(t2?.IsDefinition != false);
+
+            // We expect a given named type definition to satisfy reference identity.
+            if (t1 == t2)
+            {
+                return true;
+            }
+
+            // We relax this to permit a type to be represented by multiple symbols (e.g. separate compilations)
+            return t2 != null && t1.Name == t2.Name && t1.Arity == t2.Arity && SameOriginalSymbol(t1.ContainingSymbol, t2.ContainingSymbol);
```

Comment:
is direct name checking valid for VB cases?  i.e is case insensitivity a concern here?  Or is the expectation that that is not something between symbols.  If so, can you comment this is intentional.